### ISSUE TITLE
Require async functions to be called with the await keyword

### ISSUE THREAD

**Issue by aymericbouzy** [babel/babel-eslint#593]
Hi, I'm not sure what I'm looking for is even possible.

I often find myself in this situation : 

```js
async function foo() {
	// does some stuff
}

async function bar() {
	foo()
}
```

I have created an async function `foo` and I'm calling it some place else but I forgot the `await` keyword in front 😱 My code becomes very unpredictable and the issues I face don't help me in any way to identify where is the missing `await`.

After some time scanning my code, I finally find the culprit, and correct it:

```diff
async function foo() {
	// does some stuff
}

async function bar() {
-	foo()
+   await foo()
}
```

To be clear, this rule is different from `require-await` : the following code is not marked as invalid by the `require-await` rule : 

```js
async function foo() {}

async function bar() {
	await foo()
	foo()
}
```

Is it even possible for eslint to identify async function calls that are not awaited ?

**Comment by dnalborczyk** [comment]
@aymericbouzy   wrong project. you might wanna look here: https://github.com/eslint/eslint . babel-eslint is the underlaying "engine" (parser) of eslint.

that being said, I don't think it is possible for a variety of reasons. your "awaited" function could be an async function, or just a Promise returning function. In any case, it might be just passing through to the next function, without the Promise resolution, or you wanna assign it to a variable, and pass it elsewhere. there's no way to know what you are planning to do with the Promise.

you might wanna check out: [flow types](https://flow.org/en/). if you'd use that, flow could tell whazz up:

```javascript
async function test(): Promise<number> {
  return 1;
}

(async () => {
  const a: number = test();
})()
```

```
const a: number = test();
                       ^ Cannot assign `test()` to `a` because `Promise` [1] is incompatible with number [2].
References:
1: async function test(): Promise<number> {
                          ^ [1]
6:   const a: number = test();
              ^ [2]
```

but that might be beyond your needs. just something to keep in mind.

**Comment by aymericbouzy** [comment]
Thanks for the detailed answer! In fact I've just found this being discussed in `eslint` here : https://github.com/eslint/eslint/issues/9787 sorry ...

Using `flow` will not help me if I'm not assigning the result to any variable, though ... But it will help me most of the time.

I agree javascript allows to do all sorts of things with promises, and you've given quite a few of them. But if I want to restrict myself to (almost) always awaiting promises (except when the value is immediately returned, or passed to `Promise.all` / `Promise.race`), it could be a nice rule to have, if only it could be implemented, and you've confirmed my thinking that it couldn't be done.
